File: components/backend/installs/deps/deparse.js
const fs = require('fs');
const path = require('path');

const deps = JSON.parse(fs.readFileSync(path.join(__dirname, 'dependencies.json'), 'utf8'));

// Flatten nested object into single-level key-value map
function flatten(obj, prefix = '') {
    const result = {};
    for (const [key, value] of Object.entries(obj)) {
        const fullKey = prefix ? `${prefix}.${key}` : key;
        if (typeof value === 'object' && value !== null) {
            Object.assign(result, flatten(value, fullKey));
        } else {
            result[key] = value;
            result[fullKey] = value;
        }
    }
    return result;
}

const lookup = flatten(deps);

// Search: returns the string value for a key
function search(key) {
    return lookup[key] ?? null;
}

// Parse: resolves #-references recursively
function parse(key) {
    let value = search(key);
    if (value === null) return null;
    
    const visited = new Set([key]);
    while (typeof value === 'string' && value.startsWith('#')) {
        const refKey = value.slice(1).trim();
        if (visited.has(refKey)) return null; // circular reference guard
        visited.add(refKey);
        value = search(refKey);
        if (value === null) return null;
    }
    return value;
}

module.exports = { search, parse };

File: components/backend/installs/deps/deparse.py
#!/usr/bin/env python3
"""
deparse.py — Dependency lookup and reference resolution.

Reads dependencies.json and provides:
  - search(key): returns the string value for a key
  - parse(key): resolves #-references recursively

Usage:
    from deparse import search, parse

    value = search('torch.cpu')
    resolved = parse('torch.cpu')
"""

import json
from pathlib import Path
from typing import Optional, Any


def _load_deps() -> dict:
    """Load dependencies.json from the same directory as this file."""
    deps_path = Path(__file__).resolve().parent / 'dependencies.json'
    with open(deps_path, 'r', encoding='utf-8') as f:
        return json.load(f)


def _flatten(obj: dict, prefix: str = '') -> dict:
    """Flatten nested object into single-level key-value map."""
    result = {}
    for key, value in obj.items():
        full_key = f'{prefix}.{key}' if prefix else key
        if isinstance(value, dict):
            result.update(_flatten(value, full_key))
        else:
            result[key] = value
            result[full_key] = value
    return result


# Build the lookup table once at module load time
_lookup = _flatten(_load_deps())


def search(key: str) -> Optional[Any]:
    """Return the value for a key, or None if not found."""
    return _lookup.get(key, None)


def parse(key: str) -> Optional[Any]:
    """Resolve #-references recursively.

    If a value starts with '#', it is treated as a reference to another key.
    References are resolved recursively with circular reference protection.
    """
    value = search(key)
    if value is None:
        return None

    visited = {key}
    while isinstance(value, str) and value.startswith('#'):
        ref_key = value[1:].strip()
        if ref_key in visited:
            return None  # circular reference
        visited.add(ref_key)
        value = search(ref_key)
        if value is None:
            return None

    return value